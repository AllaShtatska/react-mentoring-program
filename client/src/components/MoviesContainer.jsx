import React from 'react';
import MoviesResults from './MoviesResults'
import movies from './movies'
import SearchContainer from './SearchContainer';
import searchFilterType from './SearchFilter/searchFilterType';
import sortType from './Sort/sortType'
import BackToSearch from './BackToSearch'
import MoviePage from './MoviePage'

class MoviesContainer extends React.Component {
    constructor() {
        super();
        this.state = {
            list: [],
            filter: searchFilterType.title,
            sortType: sortType.releaseDate,
            selectedMovie:null,
            moviesOfTheSameGenre:[],
            hasError:false
        };
    }

    componentDidMount() {
        this.setState({
            list: this.sort(movies.data, this.state.sortType)
        });
    }

    componentDidCatch(error, info){
        this.setState({hasError:true});
        console.log("I crashed: " + error +" "+info);
    }

    onSwitchFilter(type) {
        if (this.state.filter === type)
            return;

        this.setState({
                filter: type
            });
    }

    sortByReleaseDate(first, second) {
        if (new Date(first.release_date) > new Date(second.release_date)) {
            return -1;
        }

        if (new Date(first.release_date) < new Date(second.release_date)) {
            return 1;
        }

        return 0;
    }

    sortByRating(first, second) {
        return second.vote_average - first.vote_average;
    }

    sort(inMovies, sortBy) {
        switch (sortBy) {
            case sortType.rating:
                return inMovies.sort(this.sortByRating);
            case sortType.releaseDate:
            default:
                return inMovies.sort(this.sortByReleaseDate);
        }
    }

    onSwitchSortType(type) {
        if (this.state.sortType === type)
            return;

        this.setState({
            sortType: type,
            list: this.sort(this.state.list, type)
        });
    }

    onSearch(filterValue) {
        var filteredMovies = movies.data;
        if (this.state.filter === searchFilterType.title) {
            filteredMovies = movies.data.filter((movie) => movie.title.toLowerCase().includes(filterValue.toLowerCase()));
        }
        else if (filterValue.trim() !== ""){
            filteredMovies = movies.data.filter((movie) => movie.genre.map(item => item.toLowerCase()).includes(filterValue.toLowerCase()));
        }
        this.setState({
            list:this.sort(filteredMovies, this.state.sortType)
        })
    }

    onSelectMovie(movie){
        this.setState({
            selectedMovie: movie,
            moviesOfTheSameGenre:this.findMoviesOfTheSameGenre(movie)
        });
    }

    findMoviesOfTheSameGenre(selectedMovie){
        if (selectedMovie === null) return [];
        return movies.data.filter((movie) => movie.genre.map(item => item.toLowerCase()).includes(selectedMovie.genre[0].toLowerCase())&&movie.id !== selectedMovie.id);
    }

    render() {
        if (this.state.selectedMovie === null)
        {
            return (
                <div>
                    <div>
                        <SearchContainer onSearch={this.onSearch.bind(this)} filter={this.state.filter} onSwitchFilter={this.onSwitchFilter.bind(this)}/>
                    </div>
                    <div>
                        <MoviesResults list={this.state.list} sortType={this.state.sortType} onSwitchSortType={this.onSwitchSortType.bind(this)} onSelectMovie={this.onSelectMovie.bind(this)}/>
                    </div>
                </div>
            );            
        }

        return (
            <div>
                <BackToSearch onSelectMovie={this.onSelectMovie.bind(this)}/>
                <MoviePage selectedMovie={this.state.selectedMovie} moviesOfTheSameGenre={this.state.moviesOfTheSameGenre} onSelectMovie={this.onSelectMovie.bind(this)}/>
            </div>
        );
    }
}

export default MoviesContainer;