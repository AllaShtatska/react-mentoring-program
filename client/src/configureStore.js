import {createStore, applyMiddleware} from 'redux';
import createSagaMiddleware from "redux-saga";
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage' // defaults to localStorage for web and AsyncStorage for react-native

import rootReducer from './reducers';

import watchFetchMovies from './sagas/watchFetchMovies';
import watchFetchMoviesOfTheSameGenre from "./sagas/watchFetchMoviesOfTheSameGenre";

const persistConfig = {
    key: 'root',
    storage,
  }
const persistedReducer = persistReducer(persistConfig, rootReducer)

const sagaMiddleware = createSagaMiddleware();

const configuredStore = createStore(persistedReducer, applyMiddleware(sagaMiddleware));
sagaMiddleware.run(watchFetchMovies);
sagaMiddleware.run(watchFetchMoviesOfTheSameGenre);

export default () => {
    let store = configuredStore
    let persistor = persistStore(store)
    return { store, persistor }
  }
  