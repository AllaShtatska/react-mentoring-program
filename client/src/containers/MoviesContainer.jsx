import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import { errorHappened } from "../actions/errorHappened";
import { moviesFetchRequested } from "../actions/moviesFetchRequested";
import { cleanMovies } from "../actions/cleanMovies";
import MoviesPageContainer from "./MoviesPageContainer";

const mapStateToProps = state => ({
  filter: state.movies.filter,
  filterValue: state.movies.filterValue,
  sortType: state.movies.sortType,
  hasError: state.movies.hasError
});

const mapDispatchToProps = dispatch => ({
  errorHappened: () => {
    dispatch(errorHappened());
  },
  moviesFetch: searchQuery => {
    dispatch(moviesFetchRequested(searchQuery));
  },
  cleanMovies: () => {
    dispatch(cleanMovies());
  }
});

class MoviesContainer extends React.Component {
  componentWillMount() {
    this.loadMovies();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.location.key !== this.props.location.key) {
      this.loadMovies();
    }
  }

  loadMovies() {
    this.props.cleanMovies();
    if (this.props.match.params.searchQuery !== undefined) {
      this.props.moviesFetch(this.props.match.params.searchQuery);
    }
  }

  componentDidCatch(error, info) {
    this.props.errorHappened();
    console.log("I crashed: " + error + " " + info);
  }

  render() {
    if (this.props.hasError) {
      return <div>There is some error</div>;
    }

    return (
      <MoviesPageContainer
        filter={this.props.filter}
        sortType={this.props.sortType}
        filterValue={this.props.filterValue}
      />
    );
  }
}

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(MoviesContainer)
);
