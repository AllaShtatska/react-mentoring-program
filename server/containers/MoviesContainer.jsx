import React from "react";
import { connect } from "react-redux";
import { withRouter } from "next/router";

import { errorHappened } from "../actions/errorHappened";
import { moviesFetchRequested } from "../actions/moviesFetchRequested";
import { cleanMovies } from "../actions/cleanMovies";
import MoviesPageContainer from "./MoviesPageContainer";

const mapStateToProps = state => ({
  filter: state.movies.filter,
  filterValue: state.movies.filterValue,
  sortType: state.movies.sortType,
  hasError: state.movies.hasError
});

const mapDispatchToProps = dispatch => ({
  errorHappened: () => {
    dispatch(errorHappened());
  },
  moviesFetch: searchQuery => {
    dispatch(moviesFetchRequested(searchQuery));
  },
  cleanMovies: () => {
    dispatch(cleanMovies());
  }
});

class MoviesContainer extends React.Component {
  componentDidCatch(error, info) {
    this.props.errorHappened();
    console.log("I crashed: " + error + " " + info);
  }

  render() {
    if (this.props.hasError) {
      return <div>There is some error</div>;
    }

    console.log(this.props);
    console.log(this.props.filter);
    console.log(this.props.filterValue);

    return (
      <MoviesPageContainer
        filter={this.props.filter}
        sortType={this.props.sortType}
        filterValue={this.props.filterValue}
      />
    );
  }
}

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(MoviesContainer)
);
