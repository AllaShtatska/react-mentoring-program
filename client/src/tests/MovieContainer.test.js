import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { PersistGate } from 'redux-persist/integration/react'
import {Provider} from "react-redux";

import MoviesContainer from '../containers/MoviesContainer';

import configureStore from "../configureStore";

const store = configureStore();

it ('renders correctly', ()=>{
    const component = renderer.create(<Provider store={store.store}>
        <PersistGate loading={null} persistor={store.persistor}>
            <MoviesContainer />
        </PersistGate>
    </Provider>);

    var jsonComponent = component.toJSON();
    expect(jsonComponent).toMatchSnapshot();
});