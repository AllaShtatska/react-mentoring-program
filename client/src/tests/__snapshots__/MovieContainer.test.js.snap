// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div>
    <div>
      <div>
        Find your movie
      </div>
      <div>
        <input
          onKeyDown={[Function]}
          type="text"
        />
      </div>
      <div>
        <div
          className="inline"
        >
          Search by
          <div
            className="inline"
          >
            <button
              className="filter-button-selected"
              onClick={[Function]}
            >
              Title
            </button>
          </div>
          <div
            className="inline"
          >
            <button
              className="filter-button-deselected"
              onClick={[Function]}
            >
              Genre
            </button>
          </div>
        </div>
        <button
          id="search"
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </div>
  </div>
  <div>
    <div>
      <div>
        <div>
          No films found
        </div>
      </div>
    </div>
  </div>
</div>
`;
